src/libraries/System.Private.Xml/src/System/Xml/XPath/Internal/XPathScanner.cs

```diff
@@ -312,23 +326,26 @@ private string ScanString()
 
         private string ScanName()
         {
-            Debug.Assert(XmlCharType.IsStartNCNameSingleChar(CurrentChar));
-            int start = _xpathExprIndex - 1;
-            int len = 0;
+            ReadOnlySpan<char> span = _xpathExpr.AsSpan(_xpathExprIndex - 1);
 
-            while (true)
+            Debug.Assert(!span.IsEmpty);
+            Debug.Assert(span[0] == CurrentChar);
+            Debug.Assert(XmlCharType.IsStartNCNameSingleChar(span[0]));
+            Debug.Assert(XmlCharType.IsNCNameSingleChar(span[0]));
+
+            int i;
+            for (i = 1; i < span.Length && XmlCharType.IsNCNameSingleChar(span[i]); i++);
+
+            if ((uint)i < (uint)span.Length)
             {
-                if (XmlCharType.IsNCNameSingleChar(CurrentChar))
-                {
-                    NextChar();
-                    len++;
-                }
-                else
-                {
-                    break;
-                }
+                _currentChar = span[i];
+                _xpathExprIndex += i;
+                return span.Slice(0, i).ToString();
             }
-            return _xpathExpr.Substring(start, len);
+
+            _currentChar = '\0';
+            _xpathExprIndex += i - 1;
```

Comment:
why `-1`?

Reply:
To match the behavior that would be there had NextChar been used.  As a simple example, let's say _xpathExpr is "hello", _currentChar is 'h', and _xpathExprIndex is 1.  With NextChar(), we would have called it 5 times, each time incrementing _xpathExprIndex if it was still less than _xpathExpr.Length, causing it to end up equal to _xpathExpr.Length, so 5.  Now here, the span has a length of 5, which means i is 5, and _xpathExprIndex + 5 would be 6.  The -1 gets it back to where it was supposed to be, basically accounting for the -1 used at the beginning to offset the span to start at the current char rather than at _xpathExprIndex, which was already moved forward.  Now in reality, this won't matter, because the only actual checks for _xpathExprIndex are for whether it's < the length, but I wanted to make localized changes that didn't make any such assumptions.